File: tests/r_parity/32_rif.R
# StatsPAI RIF/UQR decomposition parity (R side) -- Module 32.
#
# Reads data/32_rif.csv and runs the RIF approach manually using
# dineq::rif (compute the recentered influence function for the
# median) followed by an OLS-style decomposition.
# Tolerance: rel < 1e-6 on the total gap and detailed contributions.
#
# Note: dineq's `rifr` returns the RIF regression coefficients only
# (no built-in decomposition wrapper). We compute the OLS-style
# decomposition by hand to match sp's output structure.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(dineq)
})

MODULE <- "32_rif"

df <- read_csv_strict(MODULE)

# Compute RIF for each group's outcome at the median.
group0 <- df[df$female == 0, ]
group1 <- df[df$female == 1, ]

rif0 <- dineq::rif(group0$log_wage, weights = NULL, method = "quantile",
                    quantile = 0.5)
rif1 <- dineq::rif(group1$log_wage, weights = NULL, method = "quantile",
                    quantile = 0.5)

# OLS regression of RIF on covariates within each group.
fit0 <- lm(rif0 ~ educ + exper, data = group0)
fit1 <- lm(rif1 ~ educ + exper, data = group1)

# Means of covariates per group.
X0 <- as.matrix(cbind(1, group0[, c("educ", "exper")]))
X1 <- as.matrix(cbind(1, group1[, c("educ", "exper")]))
colnames(X0)[1] <- colnames(X1)[1] <- "(Intercept)"

mean_X0 <- colMeans(X0)
mean_X1 <- colMeans(X1)

# Coefficients (Group 0 = reference).
beta0 <- coef(fit0)
beta1 <- coef(fit1)

# Total gap = sp's group-1 minus group-0 means of RIF.
total_diff <- mean(rif1) - mean(rif0)

# Decomposition: explained (composition) = (X1 - X0) * beta0;
# unexplained (structure) = X1 * (beta1 - beta0).
explained <- sum((mean_X1 - mean_X0) * beta0)
unexplained <- total_diff - explained

# Per-variable explained contributions
detailed_var <- (mean_X1 - mean_X0) * beta0

rows <- list(
  parity_row(MODULE, "total_diff",  estimate = total_diff,  n = nrow(df)),
  parity_row(MODULE, "explained",   estimate = explained,   n = nrow(df)),
  parity_row(MODULE, "unexplained", estimate = unexplained, n = nrow(df))
)

for (v in c("(Intercept)", "educ", "exper")) {
  vname <- if (v == "(Intercept)") "Intercept" else v
  rows[[length(rows) + 1L]] <- parity_row(
    MODULE, paste0("explained_", vname),
    estimate = unname(detailed_var[v]), n = nrow(df))
}

write_results(MODULE, rows,
              extra = list(statistic = "median",
                           method = "manual RIF + OLS decomposition"))

File: tests/r_parity/32_rif.py
"""StatsPAI RIF/UQR decomposition parity (Python side) -- Module 32.

Generates a deterministic wage-gap dataset and runs sp.decomposition
.rif_decomposition at the median (tau=0.5). The companion 32_rif.R
uses dineq::rifr.

Tolerance: rel < 1e-6 on the total_diff and on each detailed
contribution after matching dineq's RIF quantile and R stats::density
binned-kernel convention.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "32_rif"


def make_data(n: int = 1000, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    female = rng.binomial(1, 0.5, n)
    educ = rng.normal(15, 3, n)
    exper = rng.normal(10, 5, n) - 1.0 * (female == 1)
    log_wage = (
        2.0 + 0.07 * educ + 0.02 * exper - 0.20 * female
        + rng.normal(0, 0.3, n)
    )
    return pd.DataFrame({
        "log_wage": log_wage, "educ": educ, "exper": exper,
        "female": female,
    })


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    res = sp.decomposition.rif_decomposition(
        "log_wage ~ educ + exper",
        data=df, group="female",
        statistic="quantile", tau=0.5,
        quantile_convention="dineq",
    )

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "total_diff",
                     estimate=float(res.total_diff), n=int(len(df))),
        ParityRecord(MODULE, "py", "explained",
                     estimate=float(res.explained), n=int(len(df))),
        ParityRecord(MODULE, "py", "unexplained",
                     estimate=float(res.unexplained), n=int(len(df))),
    ]

    for _, row in res.detailed.iterrows():
        v = row["variable"]
        rows.append(ParityRecord(
            MODULE, "py", f"explained_{v}",
            estimate=float(row["explained"]), n=int(len(df))))

    write_results(
        MODULE, "py", rows,
        extra={
            "statistic": "quantile",
            "tau": 0.5,
            "quantile_convention": "dineq",
            "reference_backend_note": (
                "Track-A parity uses quantile_convention='dineq', which "
                "mirrors R dineq::rif (Hmisc type-7 quantile, R "
                "stats::density binned Gaussian density, and I(y < q))."
            ),
        },
    )


if __name__ == "__main__":
    main()
